src/wasm2asm.h

```diff
@@ -1433,6 +1448,155 @@ Ref Wasm2AsmBuilder::processFunctionBody(Function* func, IString result) {
+  Ref call = ValueBuilder::makeCall(IString(ASM_FUNC), lib, env, mem);
+  Ref ret = ValueBuilder::makeVar();
+  ValueBuilder::appendToVar(ret, ASM_MODULE, call);
+  return ret;
+}
+
+static void prefixCalls(Ref asmjs) {
+  if (asmjs->isArray()) {
+    ArrayStorage& arr = asmjs->getArray();
+    for (Ref& r : arr) {
+      prefixCalls(r);
+    }
+    if (arr.size() > 0 && arr[0]->isString() && arr[0]->getIString() == CALL) {
+      assert(arr.size() >= 2);
+      Ref prefixed = ValueBuilder::makeDot(ValueBuilder::makeName(ASM_MODULE),
+                                           arr[1]->getIString());
+      arr[1]->setArray(prefixed->getArray());
+    }
+  }
+}
+
+Ref Wasm2AsmBuilder::makeAssertReturnFunc(SExpressionWasmBuilder& sexpBuilder,
+                                          Builder& wasmBuilder,
+                                          Element& e, Name testFuncName) {
+  Expression* actual = sexpBuilder.parseExpression(e[1]);
+  Expression* expected = sexpBuilder.parseExpression(e[2]);
+  WasmType resType = expected->type;
+  actual->type = resType;
+  BinaryOp eqOp;
+  switch (resType) {
+    case i32: eqOp = EqInt32; break;
+    case i64: eqOp = EqInt64; break;
+    case f32: eqOp = EqFloat32; break;
+    case f64: eqOp = EqFloat64; break;
+    default:
+      std::cerr << "Unhandled type in assert: " << resType << std::endl;
+      abort();
+  }
+  Binary* test = wasmBuilder.makeBinary(eqOp, actual, expected);
+  std::unique_ptr<Function> testFunc(
+    wasmBuilder.makeFunction(testFuncName,
+                             std::vector<NameType>{},
+                             i32,
+                             std::vector<NameType>{},
+                             test)
+  );
+  Ref jsFunc = processFunction(testFunc.get());
+  prefixCalls(jsFunc);
+  return jsFunc;
+}
+
+Ref Wasm2AsmBuilder::makeAssertTrapFunc(SExpressionWasmBuilder& sexpBuilder,
+                                        Builder& wasmBuilder,
+                                        Element& e, Name testFuncName) {
+  Name innerFuncName("f");
+  Expression* expr = sexpBuilder.parseExpression(e[1]);
+  std::unique_ptr<Function> exprFunc(
+    wasmBuilder.makeFunction(innerFuncName,
+                             std::vector<NameType>{},
+                             expr->type,
+                             std::vector<NameType>{},
+                             expr)
+  );
+  IString expectedErr = e[2]->str();
+  Ref innerFunc = processFunction(exprFunc.get());
+  Ref outerFunc = ValueBuilder::makeFunction(testFuncName);
+  outerFunc[3]->push_back(innerFunc);
+  Ref tryBlock = ValueBuilder::makeBlock();
+  ValueBuilder::appendToBlock(tryBlock, ValueBuilder::makeCall(innerFuncName));
+  Ref catchBlock = ValueBuilder::makeBlock();
+  ValueBuilder::appendToBlock(
+    catchBlock,
+    ValueBuilder::makeReturn(
+      ValueBuilder::makeCall(
+        ValueBuilder::makeDot(
+          ValueBuilder::makeDot(
+            ValueBuilder::makeName((IString("e"))),
```

Comment:
This is kind of convoluted to read (`call(dot(dot('e', 'message'), 'includes'), expectedErr)`)
The dot-expressions feel like one logical unit, so that's harder to mentally map to and from an AST I feel.

I think this is a good time to extract `e.message.includes` to a variable.
Or to add a vararg helper method, `makeDotted("e", "message", "includes")`. (Actually kidding, but I'm not opposed)